### PR TITLE
west: add the hal_wch

### PR DESCRIPTION
**Pull request by kholia**
The hal_wch is based on the https://github.com/cnlohr/ch32v003fun work.

Context:

- https://github.com/zephyrproject-rtos/zephyr/issues/76261
- https://github.com/zephyrproject-rtos/zephyr/issues/76261#issuecomment-2260101347
- https://github.com/zephyrproject-rtos/zephyr/pull/73761

I believe we need to host this `hal_wch` in the `https://github.com/zephyrproject-rtos/` namespace at some point.

CC @nzmichaelh @fabiobaltieri @cnlohr.

### CODE REVIEW

File: west.yml
Comment (fabiobaltieri) [inline west.yml:27]: Could you send https://github.com/kholia/hal_wch/commit/787d57f0effc84205d2ea10489b7fe7cd1c4cded up to https://github.com/cnlohr/ch32v003fun and then use https://github.com/cnlohr/ch32v003fun directly here? Sounds like @cnlohr is happy to have that file there. The rest looks good, I'll rework the proposal and bring it up to the TSC.

**Comment by fabiobaltieri** [comment]
Can you also add an entry in `MAINTAINERS.yml`? See https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L4495 for example, I'd imagine with you and @nzmichaelh as maintainer, at least initially.

File: west.yml
Comment (kholia) [reply west.yml:27]: Sure. https://github.com/cnlohr/ch32v003fun/pull/364 is open now for @cnlohr's review.

**Comment by fabiobaltieri** [comment]
Brought this up at TSC today, no concerns so far, the proposal will be voted once the PR is approved.

File: west.yml
Comment (fabiobaltieri) [reply west.yml:27]: Great stuff, you can already test it on the PR by setting this to `https://github.com/cnlohr` and the one below to `ch32v003fun` and revision to `pull/364/head`.

Comment (kholia) [reply west.yml]: Done. I will re-push once the PR (https://github.com/cnlohr/ch32v003fun/pull/364) is merged.

I have also added @nzmichaelh @kholia (myself) to `MAINTAINERS.yml` file now.

**Comment by fabiobaltieri** [comment]
Ok nearly there, the system expects a file to be in `modules/hal_wch/Kconfig` now, just put a placeholder there (like https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/hal_nxp/Kconfig). Then it's also complaining about the module named `ch32v003fun`  but that's fine, it won't have that name in the fork anyway.

File: west.yml
Comment (cnlohr) [inline west.yml:255]: This is strange - do you want to pin to a specific hash or something?  Just FYI - we will likely do a proper release after the 002, 004, 005, and 006 release.

Or is this really how it's supposed to go?

File: west.yml
Comment (cnlohr) [reply west.yml]: Oh, I'm sorry, I just saw this discussion here after commenting.  That makes more sense - what is the normal process for updating to newer versions of sub repos?  Is the idea just keep it pinned until there are bugs?

Side-note: PR 346 has been accepted to 003fun.

File: west.yml
Comment (fabiobaltieri) [reply west.yml:255]: It's how we do module updates: this is just a temporary thing to have the Zephyr CI pick up the change and run on the live module PR, our automation detects this and tags the Zephyr PR as DNM automatically, but in the meantime we can validate that the CI passes with the change. Then once the module PR is merged this gets replaced with the actual SHA.

Full process: https://docs.zephyrproject.org/latest/develop/modules.html#submitting-changes-to-modules

Comment (fabiobaltieri) [reply west.yml:255]: And yeah since it's been merged: @kholia you can use `2d464950429504417a2cb33bc5313739500e518e` now, then I guess you may want to also cherry pick it into https://github.com/zephyrproject-rtos/zephyr/pull/76540 and I'll ask the TSC to create the repo next week.

Comment (cnlohr) [reply west.yml:255]: Sorry, that I can't do much else to review this PR, as I am unfamiliar with most of the inner workings of Zephyr.

Comment (fabiobaltieri) [reply west.yml:255]: That would be a bit of a nightmare :-) no it's just so we can run the full upstream CI and verify it passess before the module one is merged, which is exactly what we did (apart from the compliance error but that's just the repo name)

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by kholia** [comment]
The changes were tested with the following command:

```
$ west update
=== updating ch32v003fun (modules/hal/wch):
--- ch32v003fun: fetching, need revision 2d464950429504417a2cb33bc5313739500e518e
remote: Enumerating objects: 48, done.
remote: Counting objects: 100% (41/41), done.
remote: Compressing objects: 100% (10/10), done.
remote: Total 48 (delta 31), reused 40 (delta 31), pack-reused 7
Unpacking objects: 100% (48/48), 24.85 KiB | 591.00 KiB/s, done.
From github.com:cnlohr/ch32v003fun
 * [new branch]      add_snprintf                 -> refs/west/add_snprintf
 * [new branch]      avoid-usb-crash              -> refs/west/avoid-usb-crash
 * [new branch]      master                       -> refs/west/master
 * [new branch]      relicense                    -> refs/west/relicense
 * [new branch]      reverse_printf               -> refs/west/reverse_printf
 * [new branch]      switch_to_print_memory_usage -> refs/west/switch_to_print_memory_usage
 * [new branch]      working_on_ethernet          -> refs/west/working_on_ethernet
 * [new branch]      working_on_gdb               -> refs/west/working_on_gdb
Previous HEAD position was 04cadf3 Add support for Zephyr
```

File: MAINTAINERS.yml
Comment (kartben) [inline MAINTAINERS.yml:4635]:
```suggestion
  files:
    - modules/hal_wch/
```

File: modules/hal_wch/Kconfig
Comment (kartben) [inline modules/hal_wch/Kconfig:1]: @fabiobaltieri it looks like most HALs tend to have their Kconfig directly under modules/, in the form of Kconfig.hal_name?
Also, should this really be empty or should it define something like `HAS_WCH_HAL`?

**Review by kartben — CHANGES_REQUESTED** [review]
request for change is for the maintainers file which seems wrong with no file filter, but also a comment on location/contents of the Kconfig file. Thanks!

File: MAINTAINERS.yml
Comment (fabiobaltieri) [reply MAINTAINERS.yml:4635]: oh good point

Comment (kholia) [reply MAINTAINERS.yml]: Fixed now - thanks!

File: modules/hal_wch/Kconfig
Comment (fabiobaltieri) [reply modules/hal_wch/Kconfig:1]: Depends on how the integration gets implemented but I don't think there's the need for any Kconfig options for now. It will need a CMakeLists.txt though, something like

```cmake
if(CONFIG_SOC_SERIES_CH32V00X)

zephyr_include_directories(${ZEPHYR_HAL_WCH_MODULE_DIR}/include)

endif()
```

...or it could get a `HAS_WCH_HAL` and then that could be selected in the `soc/wch/ch32v00x/Kconfig.soc`, I think both are valid options

Comment (kholia) [reply modules/hal_wch/Kconfig:1]: I will be working on modifying the https://github.com/zephyrproject-rtos/zephyr/pull/73761 PR later today.

**Comment by kholia** [comment]
Related PR that I need help with -> https://github.com/cnlohr/ch32v003fun/pull/365

**Comment by fabiobaltieri** [comment]
> Related PR that I need help with -> [cnlohr/ch32v003fun#365](https://github.com/cnlohr/ch32v003fun/pull/365)

Sent a comment there, I think you are just missing the `modules/hal_wch/Kconfig` files on the branch where you are testing. They could either live there or in the module repo, but I think main repo is way easier to maintain.

File: modules/hal_wch/Kconfig
Comment (kholia) [reply modules/hal_wch/Kconfig:1]: PR https://github.com/zephyrproject-rtos/zephyr/pull/73761 has been updated now - cheers!

**Review by nandojve — DISMISSED** [review]

File: west.yml
Comment (fabiobaltieri) [reply west.yml:253]: the idea was to fork it to `hal_wch` in the org, same as with other hals

**Comment by nzmichaelh** [comment]
re: minichlink, I'm thinking of sending a PR to add it to the Zephyr Toolchain. Then it'll get built from source.

**Comment by fabiobaltieri** [comment]
> re: minichlink, I'm thinking of sending a PR to add it to the Zephyr Toolchain. Then it'll get built from source.

Thanks, that'd be very cool, but still having the file in that report I'm afraid means we can't just fork it in the org.

**Comment by nzmichaelh** [comment]
Understood. Pulling this PR back to the essentials and adding minichlink to the toolchain should meet the license requirements and give a good "out of the box" experience.

**Comment by fabiobaltieri** [comment]
Hey folks, update from TSC: the binary files and mixed licences in https://github.com/cnlohr/ch32v003fun unfortunately are a bit of a problem, apologies for not noticing earlier, should have gone through the repo and spot those out but, well, I didn't.

So at this point the request from the TSC is to go back to a blank `hal_wch` repo with just the essential files, which funny enough was what @kholia originally proposed if I remember right. Sorry for bring you folks to a full 360.

So next step: @kholia could you respin https://github.com/kholia/hal_wch from scratch and include only the header file, license and `module.yml`? Then update west.yml of this and https://github.com/zephyrproject-rtos/zephyr/pull/73761 with that repo and commit and next week I'll ask to get that forked in the org again.

@cnlohr at this point the [module.yml](https://github.com/cnlohr/ch32v003fun/blob/master/zephyr/module.yml) in your repo won't be used, feel free to drop it. Apologies for the noise, didn't have a full picture until now.

That's it, once that's done I'm fairly confident we can wrap the header story next week and finally let the implementation review go through.

Again, sorry for the delay and u-turns, hopefully we have a coherent picture now.

**Comment by kholia** [comment]
@fabiobaltieri Thank you!

- https://github.com/kholia/hal_wch has been recreated from scratch now
- https://github.com/zephyrproject-rtos/zephyr/pull/76540 (this PR) has been been updated
- https://github.com/zephyrproject-rtos/zephyr/pull/73761 has been updated

Thanks for the super guidance on this mission! :-)

**Comment by fabiobaltieri** [comment]
> Thanks for the superb guidance on this mission! :-)

I think that "erratic" is the word you were looking for.  :-)

File: west.yml
Comment (fabiobaltieri) [reply west.yml]: Addressed, this is now a fresh repo.

**Review by fabiobaltieri — DISMISSED** [review]
Symbolic approval, needs the module in the org.

File: MAINTAINERS.yml
Comment (kartben) [inline MAINTAINERS.yml:4715]: probably needs a `labels` entry too, and the corresponding GH label to be created too?

Comment (fabiobaltieri) [reply MAINTAINERS.yml:4715]: not necessary, the manifest action should add `manifest-hal_wch` automatically, I would not bother with an extra one, I'll create the manifest one once the repo is approved

Comment (fabiobaltieri) [reply MAINTAINERS.yml:4715]: actually not even, it's already there, guess the bot auto-created it

Comment (fabiobaltieri) [reply MAINTAINERS.yml:4715]: deleted the other one

**Review by kartben — DISMISSED** [review]

**Comment by fabiobaltieri** [comment]
No concern from the TSC but they decided to vote on this next week to let some more people check this out. Sorry this is taking forever, hopefully we are on the home stretch.

**Comment by cnlohr** [comment]
Ah! Sorry, I just saw this.  Ok! I will drop the file.

**Comment by kholia** [comment]
https://github.com/zephyrproject-rtos/hal_wch is live now - thank you!

I have updated this PR to reflect this change.

Tested using:

```
$ west update | grep -A3 hal_wch 
...
=== updating hal_wch (modules/hal/wch):
HEAD is now at 1de9d3e Add hal_wch based on the 'ch32v003fun' project
```

**Comment by fabiobaltieri** [comment]
Good job, since this is integrated in https://github.com/zephyrproject-rtos/zephyr/pull/73761 already and there's no real reason to merge it separately, let's close this one down and merge the manifest change together with https://github.com/zephyrproject-rtos/zephyr/pull/73761 whenever that's ready. Thanks for the work so far.

**Comment by alexandru0-dev** [comment]
Just a curiosity about the hal_wch.
Is it planned to expand it to other WCH MCUs or CH32v003 specific?
As personally I'm interested into the CH583 family.

I would like to contribute to the project but I have no experience with zephyr so any guidance would be more than helpful.

**Comment by kholia** [comment]
> Is it planned to expand it to other WCH MCUs or CH32v003 specific?

Yes - Support for more WCH MCUs is planned.

**Pull request closed without merging**
